application/src/main/kotlin/net/corda/v5/application/flows/FlowContextProperties.kt

```diff
@@ -0,0 +1,59 @@
+package net.corda.v5.application.flows
+
+/**
+ * Context properties of a [Flow] are key value pairs of Strings. They are comprised of two sets of properties, those
+ * set by the platform when a [Flow] is started, and those which are added by the CorDapp developer during the execution
+ * of their flow. The latter set are referred to as user context properties. All context properties are immutable once
+ * set.
+ *
+ * Where keys exist as platform properties already, setting user properties with the same key will throw. It is highly
+ * advisable to scope user property keys with some unique prefix, e.g. package name. Corda's platform keys will usually
+ * be prefixed with "corda." which is reserved, meaning that an attempt to prefix a user key with "corda." will also
+ * throw whether it exists already or not.
+ *
+ * Both sets of context properties are passed from the originating [Flow] to sub-flows, initiated flows, and services.
+ * Where sub-flows and initiated flows have extra user properties added, these are only visible in the scope of those
+ * flows and any of their sub-flows, initiated flows or services, but not back up the flow stack to any flows which
+ * sit above the one in which they were added. In this way context properties can also be thought of as a stack,
+ * unwinding along with the flow stack.
+ */
+interface FlowContextProperties {
+    /**
+     * Puts a user value into the context property store.
```

Comment:
We need to explain the scoping of the properties. i.e. when set by a flow/subflow it is available in all child subflows unless overwritten by one of them. If overwritten then the original value is restored when that child subflow returns.

Reply:
on further discussion offline it seems this is resolved by moving the existing explanation to the methods, plus adding a bit about overwrites, so that's what I've done